docs_src/design/adr/security/0026-authentication.md

```diff
@@ -0,0 +1,296 @@
+  as implemented by Hashicorp Consul, is secured over a local HTTP
+  socket with token-based authentication,
+  with the token being mediated by Hashicorp Vault.
+  An access control list limits access to the keyspace of
+  the configuration store.
+
+* Communication with EdgeX's default database, Redis, is secured using
+  username/password authentication, with the password stored
+  in Hashicorp Vault.  An access control list limits the commands
+  that clients are allowed to issue to the server.
+
+External access to EdgeX microservices has also been secured.
+EdgeX microservices only bind to local ports,
+and are only exposed externally through a Kong API gateway.
+This gateway is configured to use TLS 1.3,
+using JWT authentication using public key cryptography.
+All external requests are filtered at the API gateway.
+URL rewriting is used to concentrate microservices
+on a single HTTP-accessible port.
+
+Behind the proxy, it is not possible to verify Kong
+as the origin of local network traffic because mutual-auth TLS
+is not supported in the open source version of Kong.
+Although the Kong JWT plugin will set request headers
+on the backend request that identify the caller,
+there is no mechanism by which Kong can prove to a
+backend service that it was the component that
+performed the authentication step.
+Even though the original JWT passes through the proxy,
+the Kong authentication plugins do not expose
+token introspection endpoints that the backend service
+could use to check token validity independently.
+
+The consequence of having an API gateway that performs
+all microservice authentication is that communication
+between EdgeX microservices running behind the API gateway
+are not authenticated in any way.
+EdgeX microservices are unable to distinguish
+malicous traffic that has evaded the API gateway
+from legitimate microservice traffic.
+
+
+## Decision
+
+This ADR expands the use of Vault to include
+issuing and validating [JWTs](https://www.rfc-editor.org/rfc/rfc7519)
+for use in microservice authentication among EdgeX services.
+Vault, and not Kong, owns JWTs issuance and validation
+in the new propsal.
+
+Having been relieved of JWT management responsibility,
+it is now possible to replace the Kong API gateway and its associated
+Postgres database with a much simpler (and smaller) NGINX reverse proxy.
+All inbound requests can be authorized by an external plugin
+that can in turn delegate the JWT validity check to Vault.
+From the perspective of an external caller,
+the existing JWT authentication feature is unchanged.
+TLS termination enabled by default so as to be consistent with
+[ADR 0015 - Encryption between microservices](./0015-in-cluster-tls.md).
+
+Behind the proxy, there are two major changes:
+
+1. Every EdgeX service, when security is enabled,
+   requires a JWT be passed as part of the HTTP request
+   that is validated using Vault's token introspection endpoint.
```

Comment:
performance impact?

Reply:
TBD.  If it's too costly, it would be possible to query Vault for the validation key and validate locally.  But I suspect if the timing is that tight that request authentication blows everything out of the water, one would be running without security anyway.